### PR TITLE
Update to JDK21 - both JDK and language level

### PR DESCRIPTION
**Pull request by koppor**
Fixes https://github.com/JabRef/jabref-issue-melting-pot/issues/309

No JDK 21u1 with Adoptium as of today (https://github.com/adoptium/temurin/issues/6)

I checked available SDKs. I read through https://bell-sw.com/blog/oracle-java-alternatives-comparison-of-openjdk-distributions/#mcetoc_1h197r93f2ih and https://github.com/SAP/SapMachine/wiki/Differences-between-SapMachine-and-OpenJDK. "Liberica" came out best for me. Still using Temurin in the docs.

```cmd
koppor@DESKTOP-KAK953S:/mnt/c/Users/koppor$ sdk list java | grep 21                       
 Corretto      |     | 21           | amzn    |            | 21-amzn                      
               |     | 21.0.1       | amzn    |            | 21.0.1-amzn                  
               |     | 11.0.21      | amzn    |            | 11.0.21-amzn                 
 GraalVM CE    |     | 21           | graalce |            | 21-graalce                   
 GraalVM Oracle|     | 21           | graal   |            | 21-graal                     
               |     | 21.0.1       | graal   |            | 21.0.1-graal                 
               |     | 21           | open    |            | 21-open                      
               |     | 21.ea.35     | open    |            | 21.ea.35-open                
               |     | 21.0.1       | open    |            | 21.0.1-open                  
 Liberica      |     | 21.fx        | librca  |            | 21.fx-librca                 
               |     | 21.0.1.fx    | librca  |            | 21.0.1.fx-librca             
               |     | 21.0.1       | librca  |            | 21.0.1-librca                
               |     | 21           | librca  |            | 21-librca                    
               |     | 11.0.21.fx   | librca  |            | 11.0.21.fx-librca            
               |     | 11.0.21      | librca  |            | 11.0.21-librca               
               |     | 23.1.r21     | nik     |            | 23.1.r21-nik                 
               |     | 23.1.r21     | mandrel |            | 23.1.r21-mandrel             
 Microsoft     |     | 21           | ms      |            | 21-ms                        
 Oracle        |     | 21.0.1       | oracle  |            | 21.0.1-oracle                
               |     | 21           | oracle  |            | 21-oracle                    
 SapMachine    |     | 21           | sapmchn |            | 21-sapmchn                   
               |     | 21.0.1       | sapmchn |            | 21.0.1-sapmchn               
               |     | 11.0.21      | sapmchn |            | 11.0.21-sapmchn              
 Temurin       |     | 21           | tem     |            | 21-tem                       
               |     | 11.0.21      | tem     |            | 11.0.21-tem                  
 Zulu          |     | 21           | zulu    |            | 21-zulu                      
               |     | 21.crac      | zulu    |            | 21.crac-zulu                 
               |     | 21.fx        | zulu    |            | 21.fx-zulu                   
               |     | 21.0.1       | zulu    |            | 21.0.1-zulu                  
               |     | 11.0.21      | zulu    |            | 11.0.21-zulu               
```  

### Mandatory checks
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by koppor** [comment]
About dialog looks good:

image

File: src/main/java/org/jabref/logic/bst/BstVM.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public String render(Collection<BibEntry> bibEntries) {
         return render(bibEntries, null);
     }
 
-    protected Stack<Object> getStack() {
+    protected Deque<Object> getStack() {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/bst/BstVM.java:96]: Please revert this modernizer issue here, stack das

File: src/main/java/org/jabref/logic/util/io/FileUtil.java [hunk 2]
```diff
@@ -174,19 +171,19 @@ public static Optional<String> getUniquePathFragment(List<String> paths, Path co
      * @return the minimal unique path substring for each file path
      */
     public static List<String> uniquePathSubstrings(List<String> paths) {
-        List<Stack<String>> stackList = new ArrayList<>(paths.size());
+        List<Deque<String>> stackList = new ArrayList<>(paths.size());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/util/io/FileUtil.java:174]: Stack iteration returns in a wrong different order than deque that's why the test fails.

File: src/main/java/org/jabref/logic/bst/BstVM.java (hunk 1 above)
Comment (koppor) [reply src/main/java/org/jabref/logic/bst/BstVM.java:96]: Tests run through. Is there a special reason? I did not find the PR where you tried.

File: src/main/java/org/jabref/logic/util/io/FileUtil.java (hunk 2 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java:174]: Might have the same effect in the other places

File: src/main/java/org/jabref/logic/bst/BstVM.java (hunk 1 above)
Comment (koppor) [reply src/main/java/org/jabref/logic/bst/BstVM.java:96]: Note thar I needed to use LinkedList as implementation to allow for nulls in the stack...

File: src/main/java/org/jabref/logic/util/io/FileUtil.java (hunk 2 above)
Comment (koppor) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java:174]: > Stack iteration returns in a wrong different order than deque that's why the test fails.

According to JDK not:

image

I did not use the dequeue methods, but sticked to the stack ones:

image

image

This is so strange, I will debug now.

Comment (koppor) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java:174]: JavaDoc: https://docs.oracle.com/javase/8/docs/api/java/util/Deque.html#push-E-

Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/util/io/FileUtil.java:174]: see https://stackoverflow.com/a/54536611/3450689
https://stackoverflow.com/a/74487335/3450689

**Comment by tobiasdiez** [comment]
Good job everyone! Nice to see that the work on fixing the stupid JDK bug paid off and came to a nice conclusion.

**Review by Siedlerchr — APPROVED** [review]

**Review by calixtus — APPROVED** [review]

**Pull request merged**